diff --git a/src/dxvk/dxvk_options.cpp b/src/dxvk/dxvk_options.cpp
index 2a9f91e..06e9142 100644
--- a/src/dxvk/dxvk_options.cpp
+++ b/src/dxvk/dxvk_options.cpp
@@ -4,7 +4,7 @@ namespace dxvk {
 
   DxvkOptions::DxvkOptions(const Config& config) {
     allowMemoryOvercommit = config.getOption<bool>("dxvk.allowMemoryOvercommit", false);
-    asyncPipeCompiler     = false; // config.getOption<bool>("dxvk.asyncPipeCompiler",     false);
+    asyncPipeCompiler     = config.getOption<bool>("dxvk.asyncPipeCompiler",     false);
   }
 
 }
\ No newline at end of file
diff --git a/src/util/config/config.cpp b/src/util/config/config.cpp
index 9bbda32..73dcd72 100644
--- a/src/util/config/config.cpp
+++ b/src/util/config/config.cpp
@@ -58,6 +58,10 @@ namespace dxvk {
     { "Overwatch.exe", {{
       { "d3d11.fakeStreamOutSupport",       "True" },
     }} },
+    /* Path of Exile                              */
+    { "PathOfExile_x64.exe", {{
+      { "dxvk.asyncPipeCompiler",           "True" },
+    }} },
   }};
 
 
